File: fsharp/queen-attack/QueenAttack.fs
module QueenAttack

    let private isOnBoard (location:int): bool = location > -1 && location < 8

    let private absoluteDifference (a:int) (b:int):int =
        a - b
        |> abs

    let private absoluteDistance (row1:int, column1:int) (row2:int, column2:int):int * int =
        let rowDistance:int = absoluteDifference row1 row2
        let columnDistance:int = absoluteDifference column1 column2
        (rowDistance, columnDistance)

    let create (row:int, column: int):bool =
        let validRow:bool =
            row
            |> isOnBoard
        let validColumn:bool =
            column
            |> isOnBoard
        validRow && validColumn

    let canAttack (queen1:int * int) (queen2:int * int):bool =
        let valid1:bool =
            queen1
            |> create
        let valid2:bool =
            queen2
            |> create
        let (rowDistance, columnDistance):int * int = absoluteDistance queen1 queen2
        valid1 && valid2 && (rowDistance = columnDistance || rowDistance = 0 || columnDistance = 0)

File: solutions/scala/queen-attack/2/src/main/scala/QueenAttack.scala
case class Queen(val row:Int, val column: Int)

object Queen
{

    private def isOnBoard (location:Int):Boolean = location > -1 && location < 8

    def create(x:Int, y:Int):Option[Queen] =
    {
        def validRow:Boolean = isOnBoard(x)
        def validColumn:Boolean = isOnBoard(y)
        validRow && validColumn match
        {
            case true => Some(Queen(row = x, column = y))
            case false => None
        }
    }
}

object QueenAttack
{

    private def absoluteDifference (a:Int, b:Int):Int = Math.abs(a - b)

    private def absoluteDistance (queen1:Queen, queen2:Queen):(Int, Int) =
    {
        def rowDistance:Int = absoluteDifference(queen1.row, queen2.row)
        def columnDistance:Int = absoluteDifference(queen1.column, queen2.column)
        (rowDistance, columnDistance)
    }

    def canAttack (queen1:Queen, queen2:Queen):Boolean =
    {
        def distance:(Int, Int) = absoluteDistance(queen1, queen2)
        def rowDistance:Int = distance._1
        def columnDistance:Int = distance._2
        rowDistance == columnDistance || rowDistance == 0 || columnDistance == 0
    }

}